src/pip/_internal/commands/index.py

```diff
@@ -0,0 +1,131 @@
+        )
+
+        self.parser.insert_option_group(0, index_opts)
+        self.parser.insert_option_group(0, self.cmd_opts)
+
+    def run(self, options, args):
+        # type: (Values, List[Any]) -> int
+        handlers = {
+            "versions": self.get_available_package_versions,
+        }
+
+        # Determine action
+        if not args or args[0] not in handlers:
+            logger.error(
+                "Need an action (%s) to perform.",
+                ", ".join(sorted(handlers)),
+            )
+            return ERROR
+
+        action = args[0]
+
+        # Error handling happens here, not in the action-handlers.
+        try:
+            handlers[action](options, args[1:])
+        except PipError as e:
+            logger.error(e.args[0])
+            return ERROR
+
+        return SUCCESS
+
+    def _build_package_finder(
+        self,
+        options,  # type: Values
+        session,  # type: PipSession
+        target_python=None,  # type: Optional[TargetPython]
+        ignore_requires_python=None,  # type: Optional[bool]
+    ):
+        # type: (...) -> PackageFinder
+        """
+        Create a package finder appropriate to the index command.
+        """
+        link_collector = LinkCollector.create(session, options=options)
+
+        # Pass allow_yanked=False to ignore yanked versions.
+        selection_prefs = SelectionPreferences(
+            allow_yanked=False,
+            allow_all_prereleases=options.pre,
+            ignore_requires_python=ignore_requires_python,
+        )
+
+        return PackageFinder.create(
+            link_collector=link_collector,
+            selection_prefs=selection_prefs,
+            target_python=target_python,
+        )
+
+    def get_available_package_versions(self, options, args):
+        # type: (Values, List[Any]) -> None
+        if len(args) != 1:
+            raise CommandError('You need to specify exactly one argument')
+
+        target_python = cmdoptions.make_target_python(options)
+        query = args[0]
+
+        with self._build_session(options) as session:
+            finder = self._build_package_finder(
+                options=options,
+                session=session,
+                target_python=target_python,
+                ignore_requires_python=options.ignore_requires_python,
+            )
+            versions = set([
+                candidate.version for candidate
+                in finder.find_all_candidates(query)])
+
+            if not options.pre:
+                # Remove prereleases
+                versions = set([version for version in versions
+                                if not version.is_prerelease])
+            if not versions:
+                raise DistributionNotFound(
+                    'No matching distribution found for {}'.format(query))
+
+            formatted_versions = [str(ver) for ver in sorted(
+                versions, reverse=True)]
+            latest = str(max(versions))
```

Comment:
This can be refactored to reduce the time needed to repeatedly iterate over through things. Something like

```suggestion
            versions: Iterable[BaseVersion] = (
                candidate.version
                for candidate in finder.find_all_candidates(query)
            )
            if not options.pre:
                # Remove prereleases
                versions = (version for version in versions if not version.is_prerelease)
            versions = set(versions)

            if not versions:
                raise DistributionNotFound('No matching distribution found for {!r}'.format(query))

			formatted_versions = [str(ver) for ver in sorted(versions, reverse=True)]            
            latest = formatted_versions[0]
```

Reply:
FWIW, I noticed this too; but figured that it's easy enough to change this (and other things, like the utter lack of documentation) in a follow up PR.

Calling it experimental lets us do things like that IMO. :)